Assets/MRTK/Tools/RuntimeTools/Tools/InputFeatureUsageTool/ListInputFeatureUsages.cs

```diff
@@ -22,104 +23,141 @@ public class ListInputFeatureUsages : MonoBehaviour
 
         private void Update()
         {
-            if (listInputDevicesTextMesh == null || displayFeatureUsagesTextMeshes.Length == 0)
+            if (listInputDevicesTextMesh == null || gridObjectCollection == null || displayFeatureUsagesPrefab == null)
             {
                 return;
             }
 
 #if UNITY_2019_3_OR_NEWER
-            InputDevices.GetDevicesWithCharacteristics(InputDeviceCharacteristics.Controller, controllerInputDevices);
-            InputDevices.GetDevicesWithCharacteristics(InputDeviceCharacteristics.HandTracking, handInputDevices);
+            InputDevices.GetDevicesWithCharacteristics(InputDeviceCharacteristics.Left, leftInputDevices);
+            InputDevices.GetDevicesWithCharacteristics(InputDeviceCharacteristics.Right, rightInputDevices);
 
-            List<InputDevice> inputDevices = controllerInputDevices.Union(handInputDevices).ToList();
+            List<InputDevice> inputDevices = leftInputDevices.Union(rightInputDevices).ToList();
+            int sourceCount = inputDevices.Count;
 
-            listInputDevicesTextMesh.text = $"Detected {inputDevices.Count} input source{(inputDevices.Count > 1 ? "s:" : inputDevices.Count != 0 ? ":" : "s")}\n";
+            listInputDevicesTextMesh.text = $"Detected {sourceCount} input source{(sourceCount > 1 ? "s:" : sourceCount != 0 ? ":" : "s")}\n";
 
-            for (int i = 0; i < displayFeatureUsagesTextMeshes.Length; i++)
+            for (int i = displayFeatureUsagesTextMeshes.Count; i < sourceCount; i++)
+            {
+                displayFeatureUsagesTextMeshes.Add(Instantiate(displayFeatureUsagesPrefab, gameObject.transform).GetComponentInChildren<TextMesh>());
+                // For optimal performance, only update the collection when adding the final text panel
+                if (i == sourceCount - 1)
+                {
+                    gridObjectCollection.UpdateCollection();
+                }
+            }
+
+            for (int i = 0; i < displayFeatureUsagesTextMeshes.Count; i++)
             {
                 TextMesh textMesh = displayFeatureUsagesTextMeshes[i];
                 if (textMesh == null)
                 {
                     continue;
                 }
 
-                if (i >= inputDevices.Count)
+                if (i >= sourceCount)
                 {
-                    if (textMesh.text != string.Empty)
+                    if (textMesh.transform.parent.gameObject.activeSelf)
                     {
-                        textMesh.text = string.Empty;
+                        textMesh.transform.parent.gameObject.SetActive(false);
+                        gridObjectCollection.UpdateCollection();
                     }
                     continue;
                 }
 
+                if (!textMesh.transform.parent.gameObject.activeSelf)
+                {
+                    textMesh.transform.parent.gameObject.SetActive(true);
+                    gridObjectCollection.UpdateCollection();
```

Comment:
Based on this commenct

// For optimal performance, only update the collection when adding the final text panel

Can we just call update collection at the very end of Update()?

Reply:
Updated to set and check a flag instead of updating multiple times